client/client-lib/src/mint/mod.rs

```diff
@@ -162,22 +162,29 @@ impl<'c> MintClient<'c> {
         )
     }
 
-    fn new_note_secret(&self, dbtx: &mut DatabaseTransaction<'_>) -> DerivableSecret {
-        let new_idx = self.get_last_note_index(dbtx).next();
-        dbtx.insert_entry(&LastECashNoteIndexKey, &new_idx.as_u64())
-            .expect("DB error");
+    async fn new_note_secret(&self) -> DerivableSecret {
+        let mut new_idx;
+        loop {
```

Comment:
Would appreciate feedback on this portion. I was having issues passing a mutable reference to a database transaction through an asynchronous closure, so I removed the parameter and just created a new transaction. In my opinion, its best to have the db transaction be as tight as possible here, to minimize the change a parallel client uses the same nonce at the same time. But that makes it so this function might fail, so I needed to add the loop.

Perhaps this is ok for now, and we should add task API support here to exit cleanly instead of potentially re-trying indefinitely.

Reply:
The problem with this is that the higher level operation that wanted to use this nonce might fail, and not actually use the note which will leave holes in the nonces being actually used. Which would be fine most of the time, except it the higher level operations keep trying and trying, increasing the nonce used and then abandoning them.

The nonce recovery code will definitely support nonce holes but I would rather not have this being relied on it too much, as we have to put a limit to how many nonces ahead do we look for somewhere (like 100, or 1000?).


The contention problem is real, and it effectively force only one operation incrementing nonce at the time. I would guess that for most clients this should be totally fine, but if not - maybe we should have a more elaborated scheme, where we track both last "allocated" and "actually used" nonce as two different things, at least on some best-effort basis, so that even repeatedly failing operations don't create large nonce holes. @elsirion What do you think?

Reply:
This can lead to (more) holes in the issued token indices. Right now this only happens if a transaction fails, what should be very rare. With this change and our current approach of "if committing the DB tx fails, retry operation" it could happen that we prepare a fedimint tx, try to write it to the DB, it fails, we prepare the transaction again and use the next set of e-cash notes (since their index was already updated outside the main DB tx).

Reply:
Let's just stick with this version for now, I just noticed I'll fundamentally rebuild the client in #880 anyway.

If anyone wants spend some quality time fighting with the borrow checker: 94cf63183e544494333f0355075620f3361d0c98

Reply:
If I'm reading your guy's conversation correctly, it looks like that will make things easier. Passing database transactions through async closures is creating headaches.

Reply:
![72is21](https://user-images.githubusercontent.com/84478054/204592279-eb788139-6502-4adf-b117-ad91a44eca73.jpg)